packages/api/app/launchpadAuth.js

```diff
@@ -0,0 +1,77 @@
+
+'use strict';
+
+const launchpad = require('@cumulus/common/launchpad');
+const { RecordDoesNotExist } = require('@cumulus/common/errors');
+const { AccessToken } = require('../models');
+
+/**
+ * An express middleware that checks if an incoming express
+ * request is authenticated via Launchpad
+ *
+ * @param {Object} req - express request object
+ * @param {Object} res - express response object
+ * @param {Function} next - express middleware callback function
+ * @returns {Promise<Object>} - promise of an express response object
+ */
+async function ensureAuthorized(req, res, next) {
+  // Verify that the Authorization header was set in the request
+  const authorizationKey = req.headers.authorization;
+  if (!authorizationKey) {
+    return res.boom.unauthorized('Authorization header missing');
+  }
+  // Parse the Authorization header
+  const [scheme, token] = req.headers.authorization.trim().split(/\s+/);
+
+  // Verify that the Authorization type was "Bearer"
+  if (scheme !== 'Bearer') {
+    return res.boom.unauthorized('Authorization scheme must be Bearer');
+  }
+
+  if (!token) {
+    return res.boom.unauthorized('Missing token');
+  }
+  const access = new AccessToken();
+  let accessToken;
+  try {
+    accessToken = await access.get({ accessToken: token });
+
+    if (accessToken) {
+      const userName = accessToken.username;
+      if (Date.now() > accessToken.expirationTime) {
+        return res.boom.unauthorized('Access token has expired');
+      }
+      // Adds additional metadata that authorized endpoints can access.
+      req.authorizedMetadata = { userName };
+      return next();
+    }
+  } catch (error) {
+    if (error instanceof RecordDoesNotExist) {
+      const config = {
+        api: process.env.launchpad_api,
+        passphrase: process.env.launchpad_passphrase,
+        certificate: process.env.launchpad_certificate
+      };
+
+      const userGroup = process.env.oauth_user_group;
+      const verifyResponse = await launchpad.validateLaunchpadToken(config, token, userGroup);
+
+      if (verifyResponse.status === 'success') {
```

Comment:
have you tested that 1. token is saved to the database 2. token is valid for an hour and if the token is used within an hour, no validateLaunchpadToken is called 3. token is invalid after an hour

Reply:
1. yes, 2&3. Not exactly but I can refine,